jaxlib/pytree.cc

```diff
@@ -682,6 +682,12 @@ std::unique_ptr<PyTreeDef> PyTreeDef::Compose(const PyTreeDef& inner) const {
       out->traversal_.push_back(n);
     }
   }
+  const auto& root = traversal_.back();
+  const auto& inner_root = inner.traversal_.back();
+  auto& out_root = out->traversal_.back();
+  out_root.num_nodes = (root.num_nodes - root.num_leaves) +
```

Comment:
I can see that this was buggy before, but I am uneasy about this particular fix. First, this mutates in place the nodes. Won't this alter outer_treedef? I suspect that the rest of this module was written with the assumption that the Nodes are immutable. Second, this fix restores the invariant for `num_nodes` and `num_leaves` only for one node; shouldn't it be restored for all the interior nodes in outer_treedef? 

I am not very familiar with the internals of this module. Is it true that `num_nodes` and `num_leaves` are needed only for the root of the treedef. Maybe then it makes sense to keep these fields in the treedef, instead of the Node? That would address both issues since now we do not need to mutate Nodes.

Reply:
Re mutating in place afaik `out->traversal_.push_back(n)` in the loop above will make a copy of the node, so we're only mutating the copy.

Re restoring for all interior nodes, I think you're right. In the cases I was looking carefully at we have a very simple traversal (`kLeaf, .., kLeaf, kCustom`) but for a deeply nested tree we would need to recompute the full traversal I think. I'll take a look later this week at it, but I'm surprised our tests pass as is?

Reply:
IIRC the `num_nodes` and `num_leaves` values are to allow `Children` to reconstruct the tree structure. So you'd need to call `Compose` and then `Children` to see the bug.